src/pybind/mgr/dashboard/frontend/src/app/shared/directives/auth-storage.directive.ts

```diff
@@ -0,0 +1,93 @@
+import { Directive, Input, TemplateRef, ViewContainerRef } from '@angular/core';
+
+import _ from 'lodash';
+
+import { Permission, Permissions } from '~/app/shared/models/permissions';
+import { AuthStorageService } from '~/app/shared/services/auth-storage.service';
+
+@Directive({
+  selector: '[cdAuthStorage]'
+})
+export class AuthStorageDirective {
+  permissions: Permissions;
+
+  constructor(
+    private templateRef: TemplateRef<any>,
+    private viewContainer: ViewContainerRef,
+    private authStorageService: AuthStorageService
+  ) {}
+
+  @Input('cdAuthStorage') set cdAuthStorage(condition: string | string[] | object) {
+    this.permissions = this.authStorageService.getPermissions();
+    if (this.isAuthorized(condition)) {
+      this.viewContainer.createEmbeddedView(this.templateRef);
+    } else {
+      this.viewContainer.clear();
+    }
+  }
+
+  private isAuthorized(condition: any): boolean {
+    /*
+      single permission equivalent to check e.g osd.
+      We check if the permission key exists on the Permissions class then
+      if true We check for read permission and display the component accordingly
+    */
+
+    if (_.isString(condition)) {
+      if (_.has(this.permissions, [condition, 'read']) && this.permissions[condition]['read']) {
+        return true;
+      }
+    }
+
+    /*
+      lodash _isArray matches both objects and arrays.
+      By checking if the first index contains a value we can tell if it is truly an array or not
+      Also works for _isObject
+    */
+
+    if (_.isArray(condition) && condition[0] !== undefined) {
+      const checkedPermissions: boolean[] = [];
+
+      _.forEach(condition, (value) => {
+        if (_.has(this.permissions, [value, 'read']) && this.permissions[value]['read']) {
+          checkedPermissions.push(true);
+        } else {
+          checkedPermissions.push(false);
+        }
+      });
+
+      /*
+        check if all permissions in the checkedPermissions array are true.
+        if no false values are found then requested permissions are granted so we display the component
+      */
+      if (!checkedPermissions.includes(false)) {
+        return true;
+      }
+    }
+
+    if (_.isObject(condition) && condition[0] === undefined) {
```

Comment:
I' curious, what's the second check (`condition[0] === undefined`) for?

Reply:
I used this approach because objects and arrays both pass the _isObject check. The idea was an object has no value at index 0 so by checking for a value there we can tell if it is an object or array for sure

Reply:
Aha, but if you use the if-else if approach, and you search for arrays before than for objects, then you can assume that no array reaches the object checking, right?

Reply:
Yes, this will work great